devOpcuaSup/UaSdk/DataElementUaSdk.cpp

```diff
@@ -921,8 +928,10 @@ DataElementUaSdk::readArray (char *value, epicsUInt32 len,
                         UaVariant_to(upd->getData(), arr);
                         elemsWritten = num < arr.length() ? num : arr.length();
                         for (epicsUInt32 i = 0; i < elemsWritten; i++) {
-                            strncpy(value + i * len, UaString(arr[i]).toUtf8(), len);
-                            (value + i * len)[len - 1] = '\0';
+                            l = arrayData.XmlElementArray[i].Length;
+                            if (l >= len) l = len - 1;
+                            memcpy(value + i * len, arrayData.XmlElementArray[i].Data, l);
+                            memset(value + i * len + l, 0, len - l);
```

Comment:
Can't the memset() be done outside the loop, once?

Reply:
It clears the rest of each string element in the array, just like strncpy does, but strncpy does not like to work on unsigned char *. A single `(value + i * len)[l] = 0` would work too, but may leave old data after 0.
Of do you mean to clear the whole array before we start filling it?

Reply:
The latter.
Instead of looping over the array, first writing the string, then filling up the rest with 0,
set the whole array to 0 then loop over it filling in the strings.